cms/toolbar/toolbar.py

```diff
@@ -74,28 +76,31 @@ def structure_renderer(self):
 
     @cached_property
     def structure_mode_active(self):
-        structure = get_cms_setting('CMS_TOOLBAR_URL__BUILD')
-        return self.is_staff and structure in self.request.GET
+        try:
+            match = resolve(self.request.path_info)
+        except:
+            return False
+        return self.is_staff and match.url_name == 'cms_placeholder_render_object_structure'
 
     @cached_property
     def edit_mode_active(self):
         if not self.show_toolbar:
             return False
-        return self.structure_mode_active or self.content_mode_active
+        try:
+            match = resolve(self.request.path_info)
+        except:
+            return False
+        return self.structure_mode_active or match.url_name == 'cms_placeholder_render_object_edit'
 
     @cached_property
     def content_mode_active(self):
         if self.structure_mode_active:
             # Structure mode always takes precedence
             return False
-        return self.is_staff and self.request.session.get('cms_edit', False)
+        return self.is_staff and not self.edit_mode_active
```

Comment:
content mode is active if is not structure and the user is on edit mode.
content mode != live mode

Reply:
when going through the changes again, ``content_mode_active`` is not used anymore in ``edit_mode_active``. is it fine to delete this completely or is there a special usecase for this?

Reply:
There might be external apps using it (such as versioning :wink: ). If you're providing a different way to establish whether we're in edit, preview or structure mode then make sure that knowledge gets to us on versioning. And also probably you'll be wanting to keep all the old attributes for backwards compatibility (perhaps with deprecation warnings). But I guess that's for Paulo to decide.